tools/ctl/ide/commands/commands.go

```diff
@@ -51,21 +52,66 @@ const (
+	containerName         = "oss-rebuild-debug-temp"
 )
 
-func NewRebuildCmds(app *tview.Application, rb *rebuilder.Rebuilder, modalFn modal.Fn, butler localfiles.Butler, aiClient *genai.Client, buildDefs rebuild.LocatableAssetStore, dex rundex.Reader, benches benchmark.Repository, cmdReg *commandreg.Registry) []commandreg.RebuildCmd {
+func runLocalInput(ctx context.Context, executor *local.DockerRunExecutor, dex rundex.Reader, inp rebuild.Input) error {
+	assets := rebuild.NewFilesystemAssetStore(memfs.New())
+	h, err := executor.Start(ctx, inp, build.Options{
+		// Use a constant BuildID to make sure we overwrite the container each run
+		BuildID: containerName,
+		Resources: build.Resources{
+			AssetStore: assets,
+		},
+	})
+	if err != nil {
+		return errors.Wrap(err, "starting build")
+	}
+	go io.Copy(log.Default().Writer(), h.OutputStream())
+	res, err := h.Wait(ctx)
+	if err != nil {
+		return errors.Wrap(err, "error while waiting")
+	} else if res.Error != nil {
+		return errors.Wrap(res.Error, "build failed")
+	}
+	// TODO: How should we the artifacts? Maybe butler?
+	// TODO: Log results to dex if it's writable
+	return nil
+}
+
+func runLocal(ctx context.Context, executor *local.DockerRunExecutor, dex rundex.Reader, target rebuild.Target) error {
+	cmd := exec.CommandContext(ctx, "go", "run", "tools/ctl/ctl.go", "infer", "--ecosystem", string(target.Ecosystem), "--package", target.Package, "--version", target.Version, "--artifact", target.Artifact)
+	cmd.Stderr = log.Default().Writer()
+	out, err := cmd.Output()
+	if err != nil {
+		return errors.Wrap(err, "running inference")
+	}
+	var stratOneOf schema.StrategyOneOf
+	err = json.NewDecoder(bytes.NewBuffer(out)).Decode(&stratOneOf)
+	if err != nil {
+		log.Println("Broken strategy: " + string(out))
+		return errors.Wrap(err, "decoding strategy")
+	}
+	strat, err := stratOneOf.Strategy()
+	if err != nil {
+		return errors.Wrap(err, "decoding strategy")
+	}
+	inp := rebuild.Input{
+		Target:   target,
+		Strategy: strat,
+	}
+	return runLocalInput(ctx, executor, dex, inp)
+}
+
+func NewRebuildCmds(app *tview.Application, executor *local.DockerRunExecutor, modalFn modal.Fn, butler localfiles.Butler, aiClient *genai.Client, buildDefs rebuild.LocatableAssetStore, dex rundex.Reader, benches benchmark.Repository, cmdReg *commandreg.Registry) []commandreg.RebuildCmd {
 	return []commandreg.RebuildCmd{
 		{
 			Short: "run local",
 			Func: func(ctx context.Context, example rundex.Rebuild) {
-				rb.RunLocal(ctx, example, rebuilder.RunLocalOpts{})
-			},
-		},
-		{
-			Short: "restart && run local",
-			Func: func(ctx context.Context, example rundex.Rebuild) {
-				rb.Restart(ctx)
-				rb.RunLocal(ctx, example, rebuilder.RunLocalOpts{})
+				if err := runLocal(ctx, executor, dex, example.Target()); err != nil {
+					log.Println(err)
```

Comment:
idle thought: might be worth changing the signature to support returning an error and having the printing handled centrally. Could introduce it gradually as FuncE comparably to the cli lib pattern